File: backend/src/main/java/com/voltpay/backend/controller/PaymentRequestController.java
package com.voltpay.backend.controller;

import com.voltpay.backend.model.PaymentRequest;
import com.voltpay.backend.service.PaymentRequestService;
import com.voltpay.backend.service.UserService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@RestController
@RequestMapping("/api/requests")
public class PaymentRequestController {

    private final PaymentRequestService requestService;
    private final UserService userService;
    
    // Basic rate limiter: max 10 requests per minute per phone number
    private final Map<String, List<Long>> rateLimiter = new ConcurrentHashMap<>();

    public PaymentRequestController(PaymentRequestService requestService, UserService userService) {
        this.requestService = requestService;
        this.userService = userService;
    }

    private boolean isRateLimited(String phoneNumber) {
        long now = System.currentTimeMillis();
        List<Long> timestamps = rateLimiter.computeIfAbsent(phoneNumber, k -> new ArrayList<>());
        
        synchronized (timestamps) {
            timestamps.removeIf(t -> now - t > 60000);
            if (timestamps.size() >= 10) {
                return true;
            }
            timestamps.add(now);
        }
        return false;
    }

    @PostMapping("/create")
    public ResponseEntity<?> createRequest(@RequestHeader("Authorization") String token, 
                                           @RequestBody CreateRequestDto dto) {
        if (!userService.validateToken(dto.getRequesterPhone(), token)) {
            return ResponseEntity.status(401).body(Map.of("error", "Unauthorized"));
        }
        
        if (isRateLimited(dto.getRequesterPhone())) {
            return ResponseEntity.status(429).body(Map.of("error", "Too Many Requests. Please wait."));
        }

        PaymentRequest request = requestService.createRequest(
                dto.getRequesterPhone(), dto.getRequesterUpiId(), 
                dto.getRequesterName(), dto.getPayerPhone(), 
                dto.getAmount(), dto.getNote());
        return ResponseEntity.ok(Map.of("requestId", request.getId(), "success", true));
    }

    @GetMapping("/incoming/{phoneNumber}")
    public ResponseEntity<?> getIncoming(@RequestHeader("Authorization") String token, 
                                         @PathVariable String phoneNumber) {
        if (!userService.validateToken(phoneNumber, token)) {
            return ResponseEntity.status(401).body(Map.of("error", "Unauthorized"));
        }
        return ResponseEntity.ok(requestService.getIncomingRequests(phoneNumber));
    }

    @GetMapping("/outgoing/{phoneNumber}")
    public ResponseEntity<?> getOutgoing(@RequestHeader("Authorization") String token, 
                                         @PathVariable String phoneNumber) {
        if (!userService.validateToken(phoneNumber, token)) {
            return ResponseEntity.status(401).body(Map.of("error", "Unauthorized"));
        }
        return ResponseEntity.ok(requestService.getOutgoingRequests(phoneNumber));
    }

    @PutMapping("/{requestId}/complete")
    public ResponseEntity<?> completeRequest(@RequestHeader("Authorization") String token, 
                                             @PathVariable String requestId, 
                                             @RequestBody CompleteRequestDto dto) {
        if (!userService.validateToken(dto.getPayerPhone(), token)) {
            return ResponseEntity.status(401).body(Map.of("error", "Unauthorized"));
        }
        boolean success = requestService.completeRequest(requestId, dto.getPayerPhone());
        return ResponseEntity.ok(Map.of("success", success));
    }

    @PutMapping("/{requestId}/decline")
    public ResponseEntity<?> declineRequest(@RequestHeader("Authorization") String token, 
                                            @PathVariable String requestId, 
                                            @RequestBody CompleteRequestDto dto) {
        if (!userService.validateToken(dto.getPayerPhone(), token)) {
            return ResponseEntity.status(401).body(Map.of("error", "Unauthorized"));
        }
        boolean success = requestService.declineRequest(requestId, dto.getPayerPhone());
        return ResponseEntity.ok(Map.of("success", success));
    }

    static class CreateRequestDto {
        private String requesterPhone;
        private String requesterUpiId;
        private String requesterName;
        private String payerPhone;
        private Double amount;
        private String note;
        
        public String getRequesterPhone() { return requesterPhone; }
        public void setRequesterPhone(String requesterPhone) { this.requesterPhone = requesterPhone; }
        
        public String getRequesterUpiId() { return requesterUpiId; }
        public void setRequesterUpiId(String requesterUpiId) { this.requesterUpiId = requesterUpiId; }
        
        public String getRequesterName() { return requesterName; }
        public void setRequesterName(String requesterName) { this.requesterName = requesterName; }
        
        public String getPayerPhone() { return payerPhone; }
        public void setPayerPhone(String payerPhone) { this.payerPhone = payerPhone; }
        
        public Double getAmount() { return amount; }
        public void setAmount(Double amount) { this.amount = amount; }
        
        public String getNote() { return note; }
        public void setNote(String note) { this.note = note; }
    }

    static class CompleteRequestDto {
        private String payerPhone;
        
        public String getPayerPhone() { return payerPhone; }
        public void setPayerPhone(String payerPhone) { this.payerPhone = payerPhone; }
    }
}

File: backend/src/main/kotlin/com/voltpay/backend/controller/PaymentRequestController.kt
package com.voltpay.backend.controller

import com.voltpay.backend.service.PaymentRequestService
import com.voltpay.backend.service.UserService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/requests")
class PaymentRequestController(
    private val requestService: PaymentRequestService,
    private val userService: UserService
) {

    @PostMapping("/create")
    fun createRequest(
        @RequestHeader("Authorization") token: String,
        @RequestBody dto: CreateRequestDto
    ): ResponseEntity<*> {
        if (!userService.validateToken(dto.requesterPhone ?: "", token)) {
            return ResponseEntity.status(401).body(mapOf("error" to "Unauthorized"))
        }
        val req = requestService.createRequest(
            dto.requesterPhone ?: "",
            dto.requesterUpiId,
            dto.payerPhone ?: "",
            dto.amount ?: 0.0,
            dto.note
        )
        return ResponseEntity.ok(mapOf("success" to true, "requestId" to req.id))
    }

    @GetMapping("/pending/{payerPhone}")
    fun getPendingRequests(
        @RequestHeader("Authorization") token: String,
        @PathVariable payerPhone: String
    ): ResponseEntity<*> {
        if (!userService.validateToken(payerPhone, token)) {
            return ResponseEntity.status(401).body(mapOf("error" to "Unauthorized"))
        }
        return ResponseEntity.ok(requestService.getPendingRequestsForPayer(payerPhone))
    }

    @PutMapping("/{requestId}/complete")
    fun completeRequest(
        @RequestHeader("Authorization") token: String,
        @PathVariable requestId: String,
        @RequestBody dto: CompleteRequestDto
    ): ResponseEntity<*> {
        if (!userService.validateToken(dto.payerPhone ?: "", token)) {
            return ResponseEntity.status(401).body(mapOf("error" to "Unauthorized"))
        }
        val success = requestService.completeRequest(requestId, dto.payerPhone ?: "")
        return if (success) ResponseEntity.ok(mapOf("success" to true))
        else ResponseEntity.badRequest().body(mapOf("error" to "Could not complete request"))
    }

    @PutMapping("/{requestId}/decline")
    fun declineRequest(
        @RequestHeader("Authorization") token: String,
        @PathVariable requestId: String,
        @RequestBody dto: CompleteRequestDto
    ): ResponseEntity<*> {
        if (!userService.validateToken(dto.payerPhone ?: "", token)) {
            return ResponseEntity.status(401).body(mapOf("error" to "Unauthorized"))
        }
        val success = requestService.declineRequest(requestId, dto.payerPhone ?: "")
        return if (success) ResponseEntity.ok(mapOf("success" to true))
        else ResponseEntity.badRequest().body(mapOf("error" to "Could not decline request"))
    }

    data class CreateRequestDto(
        var requesterPhone: String? = null,
        var requesterUpiId: String? = null,
        var requesterName: String? = null,
        var payerPhone: String? = null,
        var amount: Double? = null,
        var note: String? = null
    )

    data class CompleteRequestDto(
        var payerPhone: String? = null
    )
}
